src/marqo/tensor_search/validation.py

```diff
@@ -341,72 +349,90 @@ def validate_id(_id: str):
+    """
     Args:
         field: the field name
         field_content: the field when it is a dict, especially used for multimodal tensor combination field
+        structured_field_type: the type of the field (only for structured index).
         is_non_tensor_field: for multimodal tensor combination field, this should be True
         mappings: a dictionary to help validate the object field content
-
+        index_model_dimensions: the dimensions of the model of the index. used to validate custom vector field.
     Returns:
-        True or raise an error
-    '''
-    if mappings is None:
-        raise InvalidArgError(
-            f"The field `{field}` contains a dictionary field content `{field_content}`."
-            f"However, the parameter `mappings` is {mappings}. Dictionary field contents are not supported in"
-            f"Marqo unless `mappings` is provided. Please change the type of field."
-            f"If you aim to use dictionary filed content as a special field,"
-            f"please check `https://docs.marqo.ai/0.0.15/Advanced-Usage/document_fields/#multimodal-combination-object` for more info.")
-
-    if field not in mappings:
-        raise InvalidArgError(
-            f"The field `{field}` contains a dictionary field content `{field_content}`."
-            f"However, this field `{field}` is not in the add_document parameter mappings `{mappings}`, which is not supported."
-            f"Please change the type of your field content."
-            f"If you aim to use dictionary filed content as a special field,"
-            f"please check `https://docs.marqo.ai/0.0.15/Advanced-Usage/document_fields/#multimodal-combination-object` for more info.")
+        Updated field_content dict or raise an error
+    """
 
-    if mappings[field]["type"] == "multimodal_combination":
-        validate_multimodal_combination(field_content, is_non_tensor_field, mappings[field])
+    # If field is declared in mappings, it overwrites the default.
+    if mappings and field in mappings:
+        if mappings[field]["type"] == FieldType.MultimodalCombination:
+            field_content = validate_multimodal_combination(field_content, is_non_tensor_field, mappings[field])
+        elif mappings[field]["type"] == FieldType.CustomVector:
+            field_content = validate_custom_vector(field_content, is_non_tensor_field, index_model_dimensions)
+        else:
+            raise InvalidArgError(
+                f"The field `{field}` is of invalid type in the `mappings` parameter. The only object field type supported "
+                f"is `custom_vector`. However, the `mappings` provided is: {mappings}. Please change the "
+                f"type of {field}. "
+                f"See `{marqo_docs.mappings()}` for more info on object fields. "
+            )
+    else:
+        # Use type declared in structured_field_type if not in mappings
+        if structured_field_type:
```

Comment:
We're trying to move validation to individual modules. Is it possible to do this in VespaIndex classes? I'm guessing not because add docs has quite a bit of logic in tensor search?

Reply:
Yep, it'd require a bunch more refactoring, as it mixes with a lot of logic with tensor search.